File: ZeroOneMatrixBFS.java
// Time Complexity : O(m*n)
// Space Complexity : O(m*n) - new matrix and queue too
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Start with all 0's cell. Proceed level by level (immediate neighbours and update count based on level)

//This is BFS solution.

//Optimal - O(m*n)
class Solution {
    public int[][] updateMatrix(int[][] mat) {

        if(mat == null || mat.length ==0) return mat;

        int m = mat.length;
        int n = mat[0].length;
        Queue<int []> q = new LinkedList<>();

        int[][] dirs = {{1,0}, {0,-1}, {-1,0}, {0,1}}; //4 directions clockwise

        for(int i =0; i<m ;i++){
            for(int j=0; j<n;j++){
                if(mat[i][j] == 0){
                    q.add(new int[]{i,j}); // level-0
                }else{
                    mat[i][j] = -1;
                }
            }
        }
        
        int level = 0;
        int size = 0;
        while(!q.isEmpty()){
            //Process level by level
            size = q.size();
            level++; //Before processing each level increment its count. This is for distance
            while(size>0){
                int[] curr = q.poll();

                //find neighbours
                for(int[] dir : dirs){
                    int nr = curr[0] + dir[0];
                    int nc = curr[1] + dir[1];

                    //bounds check
                    if(nr>=0 && nc>=0 && nr<m && nc<n && mat[nr][nc]==-1){
                        mat[nr][nc] = level;
                        q.add(new int[]{nr,nc});
                    }
                }
                size--;
            }
        }
        return mat;
    }
}

File: ZeroOneMatrixBFS.py
"""
Time Complexity : O(mxn) m is no. of row and n is no. of cols
Space Complexity : O(mxn) m is no. of row and n is no. of cols
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""
from collections import deque
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        if len(mat) == 0:
            return mat
        m = len(mat)
        n = len(mat[0])
        dirs = [[1,0], [0,1], [-1,0], [0,-1]]
        q = deque()
        # Find all 0 and add it ot he queue and turn all 1 to -1
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    q.append([i,j])
                else:
                    mat[i][j] = -1
        # Traverse through the queue
        while q:
            size = len(q)
            # Maintain size varibale to know the levels
            for i in range(size):
                curr = q.popleft()
                # Find the neighbor rows and cols with the help of firs list
                for item in dirs:
                    nr = curr[0] + item[0]
                    nc = curr[1] + item[1]
                    # If the new row and cols are within bounds and equals -1 then 
                    # append it to the queue and increment the level or distance
                    if nr >= 0 and nr < m and nc >= 0 and nc < n and mat[nr][nc] == -1:
                        q.append([nr,nc])
                        mat[nr][nc] = mat[curr[0]][curr[1]] + 1
        return mat
                        